### PR TITLE
[mlir][gpu]add AffineScope to gpu.func op.

### PR DESCRIPTION
**Pull request by linuxlonelyeagle**
as title.
This PR in order to solve the following problem. https://github.com/llvm/llvm-project/pull/117721.
To efficiently implement the thread-to-data mapping relationship, I introduced AffineScope in gpu.func(Data or thread layout).
I think there's a lot to be gained from it.Feel free to discuss in the comments below.Thanks!

### CODE REVIEW

**Review by krzysz00 — APPROVED** [review]
Can't see any reason not to do this, approved

File: mlir/test/Dialect/Affine/ops.mlir [hunk 1]
```diff
@@ -298,3 +298,29 @@ func.func @linearize_mixed(%index0: index, %index1: index, %index2: index, %basi
...
+    }
+  }
+}
+// CHECK-SAME:        (%[[VAL_0:.*]]: memref<?x?xf32>) kernel {
+// CHECK:             %[[VAL_1:.*]] = arith.constant 1 : index
+// CHECK:             %[[VAL_2:.*]] = memref.dim %[[VAL_0]], %[[VAL_1]] : memref<?x?xf32>
+// CHECK:             %[[VAL_3:.*]] = arith.constant 0 : index
+// CHECK:             affine.for %[[VAL_4:.*]] = %[[VAL_3]] to %[[VAL_2]] step 32 {
+// CHECK:             }
+// CHECK:             gpu.return
```
Comment (ftynse) [inline mlir/test/Dialect/Affine/ops.mlir:326]: Please configure your editor to keep the trailing newline.

**Review by ftynse — APPROVED** [review]
Makes sense to me.

**Comment by linuxlonelyeagle** [comment]
Can you guys help me merge it?Thank you.

**Pull request merged**